4.1.3 pytests fail in non-systemd environments

### What happened?

Debian autopkgtests are run in an unmanaged chroot and at least on my dev notebook, the tests fail, apparently because systemd is not running:

```
Get:106 http://deb.debian.org/debian unstable/main amd64 python3-systemd amd64 235-1+b7 [43.3 kB]
[...]
Selecting previously unselected package python3-systemd.
Preparing to unpack .../91-python3-systemd_235-1+b7_amd64.deb ...
Unpacking python3-systemd (235-1+b7) ...
[...]
Setting up python3-systemd (235-1+b7) ...
[...]
=================================== FAILURES ===================================
____________________ TestPatroni.test_patroni_patroni_main _____________________

    def abstract_main(cls: Type[AbstractPatroniDaemon], configfile: str) -> None:
        """Create the main entry point of a given daemon process.

        :param cls: a class that should inherit from :class:`AbstractPatroniDaemon`.
        :param configfile:
        """
        from .config import Config, ConfigParseError
        from .log import PatroniLogger
        from .utils import parse_int

        patroni_logger = PatroniLogger()
        try:
            config = Config(configfile)
        except ConfigParseError as e:
            sys.exit(e.value)
        patroni_logger.reload_config(config.get('log', {}))

        if not __systemd_available and os.environ.get('NOTIFY_SOCKET'):
            logger.warning('Running under systemd but python-systemd package is not installed')

        thread_stack_size = None
        if 'thread_stack_size' in config:
            thread_stack_size = parse_int(config.get('thread_stack_size'), 'B')
            if thread_stack_size is None:
                logger.warning('Failed to parse thread_stack_size value "%s"', config.get('thread_stack_size')) 

        if thread_stack_size is None:
            thread_stack_size = 524288
            logger.info('Using default value thread_stack_size = %s', thread_stack_size)

        thread_stack_size = max(65536, thread_stack_size)
        try:
            stack_size(thread_stack_size)
        except Exception as e:
            logger.warning('Failed to set threading.stack_size(%s): %r', thread_stack_size, e)
        controller = cls(config, patroni_logger)
        try:
>           controller.run()

patroni/daemon.py:240:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.13/unittest/mock.py:1169: in __call__
    return self._mock_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib/python3.13/unittest/mock.py:1173: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Mock id='139689185607168'>, args = (), kwargs = {}
effect = <class 'tests.SleepException'>

    def _execute_mock_call(self, /, *args, **kwargs):
        # separate from _increment_mock_call so that awaited functions are
        # executed separately from their call, also AsyncMock overrides this method

        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
>               raise effect
E               tests.SleepException

/usr/lib/python3.13/unittest/mock.py:1228: SleepException

During handling of the above exception, another exception occurred:

self = <tests.test_patroni.TestPatroni testMethod=test_patroni_patroni_main>

    @patch('sys.argv', ['patroni.py', 'postgres0.yml'])
    @patch('time.sleep', Mock(side_effect=SleepException))
    @patch('patroni.daemon.stack_size', Mock(side_effect=Exception))
    @patch.object(etcd.Client, 'delete', Mock())
    @patch.object(AbstractEtcdClientWithFailover, '_get_machines_list', Mock(return_value=['http://remotehost:2379']))
    @patch.object(Thread, 'join', Mock())
    @patch.object(Postgresql, '_get_gucs', Mock(return_value={'foo': True, 'bar': True}))
    @patch.object(Postgresql, '_wait_for_connection_close', Mock())
    def test_patroni_patroni_main(self):
        with patch('subprocess.call', Mock(return_value=1)):
            with patch.object(Patroni, 'run', Mock(side_effect=SleepException)), \
                patch('patroni.daemon.__systemd_available', False), \
                patch.dict(os.environ, {'PATRONI_THREAD_STACK_SIZE': 'a',
                                        'PATRONI_THREAD_POOL_SIZE': 'a',
                                        'PATRONI_POSTGRESQL_DATA_DIR': 'data/test0',
                                        'NOTIFY_SOCKET': '/run/systemd/notify'}):
>               self.assertRaises(SleepException, _main)

tests/test_patroni.py:137:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
patroni/__main__.py:392: in main
    return patroni_main(args.configfile)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
patroni/__main__.py:264: in patroni_main
    abstract_main(Patroni, configfile)

patroni/daemon.py:244: in abstract_main
    controller.shutdown()
patroni/daemon.py:197: in shutdown
    notify_systemd("STOPPING=1")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def notify_systemd(msg: str) -> None:
>       daemon.notify(msg)  # pyright: ignore
        ^^^^^^^^^^^^^^^^^^
E       FileNotFoundError: [Errno 2] No such file or directory

patroni/daemon.py:51: FileNotFoundError
patroni/daemon.py:51: FileNotFoundError
----------------------------- Captured stderr call -----------------------------
2026-05-06 12:27:12,287 INFO: Patroni global thread_pool_size = 5
2026-05-06 12:27:12,287 DEBUG: New etcd client created for http://127.0.0.1:2379
2026-05-06 12:27:12,287 INFO: Selected new etcd server http://remotehost:2379
2026-05-06 12:27:12,289 INFO: No PostgreSQL configuration items changed, nothing to reload.
2026-05-06 12:27:12,289 INFO: REST API thread_pool_size = 5
2026-05-06 12:27:12,289 DEBUG: restapi.server_tokens is set to "original".
2026-05-06 12:27:12,295 WARNING: Running under systemd but python-systemd package is not installed
2026-05-06 12:27:12,295 WARNING: Running under systemd but python-systemd package is not installed
2026-05-06 12:27:12,295 WARNING: Failed to parse thread_stack_size value "a"
2026-05-06 12:27:12,295 WARNING: Failed to parse thread_stack_size value "a"
2026-05-06 12:27:12,295 INFO: Using default value thread_stack_size = 524288
2026-05-06 12:27:12,295 INFO: Using default value thread_stack_size = 524288
2026-05-06 12:27:12,295 WARNING: Failed to set threading.stack_size(524288): Exception()
2026-05-06 12:27:12,295 WARNING: Failed to set threading.stack_size(524288): Exception()
2026-05-06 12:27:12,295 WARNING: Failed to parse thread_pool_size value "a": ValueError("invalid literal for int() with base 10: 'a'")
2026-05-06 12:27:12,295 WARNING: Failed to parse thread_pool_size value "a": ValueError("invalid literal for int() with base 10: 'a'")
2026-05-06 12:27:12,295 INFO: Patroni global thread_pool_size = 5
2026-05-06 12:27:12,295 INFO: Patroni global thread_pool_size = 5
2026-05-06 12:27:12,298 INFO: Selected new etcd server http://remotehost:2379
2026-05-06 12:27:12,298 INFO: Selected new etcd server http://remotehost:2379
2026-05-06 12:27:12,299 INFO: No PostgreSQL configuration items changed, nothing to reload.
2026-05-06 12:27:12,299 INFO: No PostgreSQL configuration items changed, nothing to reload.
2026-05-06 12:27:12,300 INFO: REST API thread_pool_size = 5
2026-05-06 12:27:12,300 INFO: REST API thread_pool_size = 5
```

Not sure about those `Failed to parse thread_pool_size value` warnings.

There's no `/run/systemd/notify` in the autopkgtest environment.

### How can we reproduce it (as minimally and precisely as possible)?

Run `python3 -m pytest tests` in a non-systemd chroot

### What did you expect to happen?

Tests pass

### Patroni/PostgreSQL/DCS version

-   Patroni version: 4.1.3
-   PostgreSQL version: n/a
-   DCS (and its version): n/a


### Patroni configuration file

```yaml
n/a
```

### patronictl show-config

```yaml
n/a
```

### Patroni log files

```shell
see above
```

### PostgreSQL log files

```shell
n/a
```

### Have you tried to use GitHub issue search?

- [x] Yes

### Anything else we need to know?

_No response_

--- a/patroni/daemon.py
+++ b/patroni/daemon.py
@@ -44,18 +44,20 @@ if TYPE_CHECKING:  # pragma: no cover
 
 logger = logging.getLogger(__name__)
 
-try:  # pragma: no cover
-    from systemd import daemon  # pyright: ignore
+SOCKET_ENV_VARIABLE = 'NOTIFY_SOCKET'
+__systemd_available = False
 
-    def notify_systemd(msg: str) -> None:
-        daemon.notify(msg)  # pyright: ignore
-
-    __systemd_available = True
-except ImportError:  # pragma: no cover
-    def notify_systemd(msg: str) -> None:
+if os.environ.get(SOCKET_ENV_VARIABLE) is not None:
+    try:  # pragma: no cover
+        from systemd import daemon  # pyright: ignore
+        __systemd_available = True
+    except ImportError:  # pragma: no cover
         pass
 
-    __systemd_available = False
+
+def notify_systemd(msg: str) -> None:
+    if __systemd_available:  # pragma: no cover
+        daemon.notify(msg)  # pyright: ignore
 
 
 def get_base_arg_parser() -> argparse.ArgumentParser:
@@ -216,7 +218,7 @@ def abstract_main(cls: Type[AbstractPatroniDaemon], configfile: str) -> None:
         sys.exit(e.value)
     patroni_logger.reload_config(config.get('log', {}))
 
-    if not __systemd_available and os.environ.get('NOTIFY_SOCKET'):
+    if not __systemd_available and os.environ.get(SOCKET_ENV_VARIABLE):
         logger.warning('Running under systemd but python-systemd package is not installed')
 
     thread_stack_size = None

--- a/patroni/postgresql/postmaster.py
+++ b/patroni/postgresql/postmaster.py
@@ -13,6 +13,7 @@ from typing import Dict, List, Optional
 import psutil
 
 from patroni import KUBERNETES_ENV_PREFIX, PATRONI_ENV_PREFIX
+from patroni.daemon import SOCKET_ENV_VARIABLE
 
 # avoid spawning the resource tracker process
 if sys.version_info >= (3, 8):  # pragma: no cover
@@ -294,7 +295,7 @@ class PostmasterProcess(psutil.Process):
         # On Windows, in order to run a side-by-side assembly the specified env must include a valid SYSTEMROOT.
         # We also remove NOTIFY_SOCKET environment variable so that PostgreSQL doesn't send READY=1 or STOPPING=1
         # to systemd, because it is exclusively responsibility of Patroni to do it.
-        env = {p: os.environ[p] for p in os.environ if p != 'NOTIFY_SOCKET'
+        env = {p: os.environ[p] for p in os.environ if p != SOCKET_ENV_VARIABLE
                and not p.startswith(PATRONI_ENV_PREFIX) and not p.startswith(KUBERNETES_ENV_PREFIX)}
         if 'PG_MALLOC_ARENA_MAX' in env:
             env['MALLOC_ARENA_MAX'] = env.pop('PG_MALLOC_ARENA_MAX')

